src/analysis/complex/exponential.lean

```diff
@@ -26,93 +27,201 @@ The following functions are shown to be continuous:
+  * complex and real exponential function
+  * sin, cos, sinh, cosh
+
 ## Tags
 
 exp, log, sin, cos, tan, arcsin, arccos, arctan, angle, argument, power, square root,
 
 -/
+noncomputable theory
 
-open finset filter metric
+open finset filter metric asymptotics
 open_locale topological_space
 
 namespace complex
 
-lemma tendsto_exp_zero_one : tendsto exp (𝓝 0) (𝓝 1) :=
-tendsto_nhds_nhds.2 $ λ ε ε0,
-  ⟨min (ε / 2) 1, lt_min (div_pos ε0 (by norm_num)) (by norm_num),
-    λ x h, have h : abs x < min (ε / 2) 1, by simpa [dist_eq] using h,
-      calc abs (exp x - 1) ≤ 2 * abs x : abs_exp_sub_one_le
-          (le_trans (le_of_lt h) (min_le_right _ _))
-        ... = abs x + abs x : two_mul (abs x)
-        ... < ε / 2 + ε / 2 : add_lt_add
-          (lt_of_lt_of_le h (min_le_left _ _)) (lt_of_lt_of_le h (min_le_left _ _))
-        ... = ε : by rw add_halves⟩
+/-- The complex exponential is everywhere differentiable, with the derivative `exp x`. -/
+lemma has_deriv_at_exp (x : ℂ) : has_deriv_at exp (exp x) x :=
+begin
+  rw has_deriv_at_iff_is_o_nhds_zero,
+  have : (1 : ℕ) < 2 := by norm_num,
+  apply is_O.trans_is_o _ (is_o_pow_id this),
+  apply is_O_iff.2 ⟨∥exp x∥, _⟩,
+  have : metric.ball (0 : ℂ) 1 ∈ nhds (0 : ℂ) :=
+    mem_nhds_sets metric.is_open_ball (by simp [zero_lt_one]),
+  apply filter.mem_sets_of_superset this (λz hz, _),
+  simp only [metric.mem_ball, dist_zero_right] at hz,
+  simp only [exp_zero, mul_one, one_mul, add_comm, normed_field.norm_pow,
+             zero_add, set.mem_set_of_eq],
+  calc ∥exp (x + z) - exp x - z * exp x∥
+    = ∥exp x * (exp z - 1 - z)∥ : by { congr, rw [exp_add], ring }
+    ... = ∥exp x∥ * ∥exp z - 1 - z∥ : normed_field.norm_mul _ _
+    ... ≤ ∥exp x∥ * ∥z∥^2 :
+      mul_le_mul_of_nonneg_left (abs_exp_sub_one_sub_id_le (le_of_lt hz)) (norm_nonneg _)
+end
+
+lemma differentiable_exp : differentiable ℂ exp :=
+λx, (has_deriv_at_exp x).differentiable_at
+
+@[simp] lemma deriv_exp {x : ℂ} : deriv exp x = exp x :=
+(has_deriv_at_exp x).deriv
 
 lemma continuous_exp : continuous exp :=
-continuous_iff_continuous_at.2 (λ x,
-  have H1 : tendsto (λ h, exp (x + h)) (𝓝 0) (𝓝 (exp x)),
-    by simpa [exp_add] using tendsto_mul tendsto_const_nhds tendsto_exp_zero_one,
-  have H2 : tendsto (λ y, y - x) (𝓝 x) (𝓝 (x - x)) :=
-     tendsto_sub tendsto_id (@tendsto_const_nhds _ _ _ x _),
-  suffices tendsto ((λ h, exp (x + h)) ∘
-      (λ y, id y - (λ z, x) y)) (𝓝 x) (𝓝 (exp x)),
-    by simp only [function.comp, add_sub_cancel'_right, id.def] at this;
-      exact this,
-  tendsto.comp H1 (by rw [sub_self] at H2; exact H2))
+differentiable_exp.continuous
+
+/-- The complex sine function is everywhere differentiable, with the derivative `cos x`. -/
+lemma has_deriv_at_sin (x : ℂ) : has_deriv_at sin (cos x) x :=
+begin
+  have A : has_deriv_at (λ(z:ℂ), exp (z * I)) (I * exp (x * I)) x,
+  { have := has_deriv_at.mul (has_deriv_at_id x) (has_deriv_at_const x I),
+    simp at this,
```

Comment:
The nonterminal simp can be avoided with 
```suggestion
    refine has_deriv_at.comp x (has_deriv_at_exp _) _,
    simpa
```

Reply:
I have found that an idiom based on convert is more efficient in derivative computations, so I have modified all proofs accordingly.